File: C++/min-stack.cpp
class MinStack {
    vector<int> m;
    vector<int> v;
    int minVal = INT32_MAX;
    
public:
    MinStack() {
        
    }
    
    void push(int val) {
        m.push_back(minVal);
        minVal = min(val, minVal);
        v.push_back(val);
    }
    
    void pop() {
        minVal = m.back();
        m.pop_back();
        v.pop_back();
    }
    
    int top() {
        return v.back();
    }
    
    int getMin() {
        return minVal;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: Kotlin/min-stack.kt
class MinStack() {

    val stackVal = ArrayDeque<Int>()
    val stackMin = ArrayDeque<Int>()

    fun push(`val`: Int) {
        stackVal.push(`val`)
        val currMin = stackMin.peek() ?: `val`
        val newMin = if (`val` < currMin) `val` else currMin
        stackMin.push(newMin)
    }

    fun pop() {
        stackVal.pop()
        stackMin.pop()
    }

    fun top(): Int {
        return stackVal.peek()
    }

    fun getMin(): Int {
        return stackMin.peek()
    }

}

/**
 * Your MinStack object will be instantiated and called as such:
 * var obj = MinStack()
 * obj.push(`val`)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */